// react-escape/dist/react-escape.js
"use strict";

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var React = require('react');
var ReactDOM = require('react-dom');
var assign = require('object-assign');

var Escape = React.createClass({
    displayName: 'Escape',

    render: function render() {
        this._nodes = this.renderLayer();
        return React.createElement('noscript', { ref: 'from' });
    },
    renderLayer: function renderLayer() {
        var _props = this.props;
        var style = _props.style;

        var props = _objectWithoutProperties(_props, ['style']);

        style = style || {};
        assign(style, styles.escapeLayer, styles[props.to]);
        return React.createElement(
            'div',
            _extends({ style: style }, props),
            this.props.children
        );
    },
    componentDidMount: function componentDidMount() {
        this.escapePoint = this.refs.from.parentNode;
        var layer = document.createElement('div');
        document.body.appendChild(layer);
        this._layer = layer;
        ReactDOM.render(this._nodes, this._layer);
    },
    componentDidUpdate: function componentDidUpdate() {
        ReactDOM.render(this._nodes, this._layer);
    },
    componentWillUnmount: function componentWillUnmount() {
        this.escapePoint = undefined;
        ReactDOM.unmountComponentAtNode(this._layer);
        this._layer.parentNode.removeChild(this._layer);
        this._layer = null;
    },
    getSize: function getSize() {
        var e = this._layer.firstChild;
        return { width: e.offsetWidth, height: e.offsetHeight };
    }
});

var styles = {
    escapeLayer: {
        position: 'absolute',
        left: 0,
        top: 0,
        right: 0,
        bottom: 0,
        zIndex: 10001,
        pointerEvents: 'none',
        overflow: 'visible'
    },
    viewport: {
        position: 'fixed'
    }
};

module.exports = Escape;
